Address type system

dr-orlovsky opened the pull request:
Major refactoring of witness program processing in address and scriptPubkeys. Makes address-related types to guarantee correctness of the witness programs.

Closes #648

Kixunil commented on src/util/address.rs:
```diff
@@ -135,7 +135,189 @@ impl From<bech32::Error> for Error {
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum SegWitInfo {
+    /// P2PKH addresses
+    PreSegWit,
```
`Option::None` would feel more idiomatic, do you have a particular reason to not use it?

dr-orlovsky replied:
Yes: ambiguity with P2SH (which may hide P2WSH/WPKH-in-P2SH), as explained in comments below. So `Option<SegWitInfo>` does not work here, we need three alternatives like "non-segwit", "maybe segwit" and "segwit"

Kixunil replied:
`None` being "non-segwit" would feel natural to me but I can understand that you can treat them as being on same level.

sanket1729 replied:
I would prefer adding APIs like `witness_version` directly to `AddressInfo`. The struct `SegwitInfo` has only one function of use that provides the witness version of the address. I don't see why we need 3 constructors for it and this amount of code bloat for it.

dr-orlovsky replied:
@sanket1729 `SegWitInfo` is needed since it is returned not only by address types, but also from Script in later commits - and you cant use `AddressInfo` for that

Kixunil commented on src/util/address.rs:
```diff
@@ -150,6 +332,30 @@ pub enum AddressType {
+impl AddressType {
+    /// Extracts known [`AddressType`] from [`AddressInfo`], if any
+    #[inline]
+    pub fn with(address_info: AddressInfo) -> Option<AddressType> {
```
I think `from_address_info` is the usual naming convention?

dr-orlovsky replied:
the usual one is `TryFrom` :)

Kixunil replied:
Yes, `TryFrom` is better for conversions returning `Result`.

sanket1729 replied:
+1 for `from_address_info` . I would find it more intuitive

dr-orlovsky replied:
Done

Kixunil commented on src/util/address.rs:
```diff
@@ -176,49 +382,207 @@ impl FromStr for AddressType {
+/// Used in enumerations on Segwit info extratible from [`Address`] and [`TxOut`]/`scriptPubkey`s.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(u8)]
+pub enum FutureWitnessVersion {
```
So this type essentially guarantees next version will cause API break in this library. I see the appeal of being super-specific with types but this feels like too much.

dr-orlovsky replied:
Not next version, but a future version after bitcoin softfork - which will be a major lib update and API break anyway.

Kixunil replied:
Yeah, that's what I meant. I don't think a new soft fork *has to be* API break. If there's something with the design of this library that causes it I'd like it to be changed.

dr-orlovsky replied:
I do not get why we may like to have softfork as a non major version. I think we need to ensure that softforks are always major releases

Kixunil replied:
To not force people to change their code when they don't absolutely require it. It's already a massive pain to keep various libs in sync. Especially if you take into account MSRVs and even system dependencies.

dr-orlovsky replied:
Accepting/supporting bitcoin softfork in any lib/project should be a conscious decision without defaulting to "auto-update"

Kixunil replied:
Auto-update does not make the code magically support a soft fork though.

apoelstra replied:
We should not require a major version bump for softforks. (Though in practice there is a good chance we will.)

dr-orlovsky replied:
Removed the type

Kixunil commented on src/util/address.rs:
```diff
@@ -176,49 +382,207 @@ impl FromStr for AddressType {
+    fn with(version: &WitnessVersion) -> Self {
+        match version {
+            WitnessVersion::V0 |
+            WitnessVersion::V1 => unreachable!("internal constructor `FutureWitnessVersion::with` was called on future witness versions"),
```
`dont_panic`/`no_panic` candidate?

dr-orlovsky replied:
Not sure I understand. If you are talking about https://docs.rs/no-panic/0.1.15/no_panic/index.html than we can't add dependencies just because of this place

Kixunil replied:
Yeah, that one. I was thinking about treating it like dev-dependency (although it can't go to `[dev-dependencies]`) and start using it at some places to prove correctness. This could be one of them.

dr-orlovsky replied:
We need @apoelstra approval on having a new dependency

apoelstra replied:
This is kinda neat but I don't think it's worth pulling in an extra dep, especially one that does such deep magic.

Kixunil replied:
Not actually that much magic, it simply puts call to non-existing function into panicking branches, name of the function being the error message.

I wonder what kind of harm do you see from having it as optional and using it in CI only?

dr-orlovsky replied:
With the removal of the `FutureWitnessVersion` this code is gone

Kixunil commented on src/util/address.rs:
```diff
@@ -176,49 +382,207 @@ impl FromStr for AddressType {
-    /// Future (unsupported) version of Witness Program
+
+    /// Version 16 of witness program, represented by 0x60 byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_16`])
```
Nice doc!

dr-orlovsky replied:
Thank you :)

Kixunil reviewed (commented):
Concept ACK

dr-orlovsky commented:
Did a big update completing this PR by providing guarantees for address-related type sanity

sanket1729 commented on src/util/address.rs:
```diff
@@ -135,7 +135,100 @@ impl From<bech32::Error> for Error {
+    /// Detects if the address can be spend only with a valid signature (i.e. not anyone can spend
+    /// address of a future witness version or deterministically unspendable)
+    #[inline]
+    pub fn requires_valid_signature(self) -> bool {
```
`requires_valid_signature` is not a good name because there may be p2sh/p2wsh addresses that don't require valid signatures to spend. How about calling this `is_known`? Note that is the same as `address_type().is_some()`. I wonder if it is worth having an API for it.

sanket1729 replied:
This is also the same as the already existing `is_standard`.

sanket1729 commented on src/util/address.rs:
```diff
@@ -135,7 +135,100 @@ impl From<bech32::Error> for Error {
+    /// Detects if the address can be proved to be unspendable under current consensus rules, which
+    /// may change only with a hard fork.
+    #[inline]
+    pub fn is_deterministically_unspendable(self) -> bool {
```
This is again misleading because there can be deterministically unspendable p2sh wrapped op_return etc. 

How about renaming the above to `is_valid_recipient_address` and this one to `is_valid_destination_address`? That way it also lets the wallet developers know how they should be using the API.

dr-orlovsky replied:
We have similary named method in `Script` and in later commits I renamed this function such that it had the same name as in `Script`

dr-orlovsky replied:
Done as you suggested. This also resolves the comment above

sanket1729 commented on src/util/address.rs:
```diff
@@ -522,35 +623,40 @@ impl Address {
+                    (WitnessVersion::V0, len) => AddressInfo::NotSpendableWitnessV0(len),
+                    (WitnessVersion::V1, 32) => AddressInfo::p2tr(),
+                    (WitnessVersion::V1, len) => AddressInfo::NonTaprootWitnessV1(len),
+                    (version, len) => AddressInfo::Future(version, len),
```
Remark: Perfect, this is clean

sanket1729 reviewed (commented):
EDIT: updated after the following discussion from #648 and reading rest of the commits.
---
Left a few comments on the first commit. Did not review all of it

My personal opinion is that this PR has too much code and additional data structures for the things it tries to achieve. I think developers using this library should only care about the following from address API:

1) Users should be able to send to future addresses.
2) Users should receive on known supported addresses.

To that end, I think we can do the following:
```rust
enum AddressInfo {
// we should send and receive to these addresses
Supported(AddressType),
// We should support sending to these, but don't receive here
Future(version, len),
}
```

This keeps the code diff small, users don't really need to know taprootv1 non-32 bytes, vs future version 2. If there is someone who is interested in implementing the differences of why something is valid vs non-valid, they can do so easily downstream. Also I don't see the point of `SegwitInfo` struct. There can be simple method to get `Option<(verion, len)>`

Curious what do other people feel about this?

EDIT: updated after the following discussion from #648 and reading rest of the commits.

Kixunil commented:
@sanket1729 I agree that detailed analysis has limited usefulness but there is at least one use case: in privacy-oriented protocols/wallets (e.g. BIP78) one needs to at least compare equality of address types to get maximum privacy. I think the current proposal goes beyond that but maybe @dr-orlovsky has some other use case in mind?

RCasatta commented on src/util/address.rs:
```diff
@@ -354,6 +746,83 @@ impl From<WitnessVersion> for opcodes::All {
+            return Err(WitnessError::NotSegwitScriptPubkey)
+        }
+        let version = WitnessVersion::from_opcode(opcodes::All::from(script[0]))?;
+        let program = script[2..].to_vec();
```
is it guaranteed script[1] is a OP_PUSHDATA using 1 byte?

dr-orlovsky replied:
The data type is removed, so the question is solved

dr-orlovsky commented:
@sanket1729 agree on replacing `SegWitInfo` with `Option<(WitnessVersion, usize)>`

dr-orlovsky commented:
@kixunil my usecase is professional software (wallets, blockchain explorers - Bitcoin Pro, MyCitadel) that provides as much detailed information as possible. Also databases indexing blockchain data for research purposes.

apoelstra commented:
I share Sanket's concerns that this PR adds a lot of data structures with unclear usecases, which will need to be adjusted in breaking ways in the future.

Wallets should be using rust-miniscript to keep track of their own addresses and related information. Block explorers etc. can use their own ad-hoc analysis of scriptPubKeys ... it is not at all obvious to me that they would want to categorize this information in the way proposed by this PR. And if they did, it is not obvious to me that this library is the place to add direct support.

Kixunil commented:
Yeah, I think having it in a separate crate with integrations would be better and closer to the idea of splitting up this crate.

dr-orlovsky commented:
@sanket1729 @apoelstra @Kixunil 

I have simplified this PR by removing `FutureWitnessVersion` and `SegWitInfo` data types. Also, I have addressed all comments and suggestions. The only thing that has left is two more variants in `AddressInfo` structure since they are needed to differentiate between valid recipient addresses and valid destination addresses (see https://github.com/rust-bitcoin/rust-bitcoin/pull/668/files#diff-c55d0186f3d5990b5e223c1a8aa1bbe60988201c452dd13b3324e892bc33119eR206-R236)

sanket1729 commented:
Sorry for bringing this up again. In my opinion, there should not be any `AddressInfo` for `NonSpendableV0`. And `NonTaprootV1` can be merged into Future enum. That would mean conversion from `AddressType` to `AddressInfo` should error instead of given it a variant of its own. Even though I have a strong preference here :) , I can go either way on `NonSpendableV0`.

At the very least, I think we should merge `NonTaprootV1`(which is just a bitcoin protocol implementation detail) into `Future`. I feel having only two types would make APIs really clean. `AddressType::is_future`, `is_known`. The value of having a simple API is a) serves good documentation and b)  someone does not have to know all the inner details of bitcoin to work on a payment processor. 

If you really want to analysis on script pubkeys, I would not be opposed to supporting it in `rust-bitcoin`. But I really think that could be a separate project.

sanket1729 commented on src/util/address.rs:
```diff
@@ -234,6 +234,42 @@ impl AddressInfo {
+
+    /// Returns whether support of the address will be possible only in a future through a softfork
+    #[inline]
+    pub fn requires_softfork(self) -> bool {
```
I don't know if this is useful but sure

sanket1729 reviewed (commented):
I think we should check if the 32-byte in p2tr is actually a correct x-only key in p2tr.

tcharding reviewed (commented):
I'm not convinced this `AddressInfo` struct is helping folks out. I see the benefit of it for internals, having all the edge cases of the various variants there in code but for users coming to the library they are going to go `Address` -> `AddressInfo` -> WTF?

We already have a bunch of methods on `Address` that give useful hints about the address. Why not make `AddressInfo` private and just add methods as they are required?

apoelstra commented on src/util/address.rs:
```diff
@@ -137,7 +137,143 @@ impl From<bech32::Error> for Error {
+    Future(WitnessVersion, /** witness program length */ usize),
+    /// Witness version is 0 and is known to consensus, but witness program has non-standard length
+    /// which is not supported by consensus rules and can't be spent.
+    NotSpendableWitnessV0(/** witness program length */ usize),
```
As others have mentioned, I find this variant very confusing. There are lots of ways to construct provably unspendable outputs, and this one is particularly obscure. It's also not clear that you could actually produce a usable address for something like this, even if you bech32-encoded it, because parsers tend to have length checks.

Similarly `NonTaprootWitnessV1` should be merged into `Future`.

apoelstra commented:
Even with the reduced scope I agree with others that this PR has an unclear motivation and purpose. My original comment, that users who want to classify addresses should do so themselves and/or use descriptors (and that if they do, it seems pretty unlikely to me that they would want this particular ad-hoc classification scheme), still stands.

dr-orlovsky commented:
Ok, closing it
